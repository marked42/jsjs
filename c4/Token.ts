export enum TokenType {
  EOF = -1,
  StringLiteral,
  Identifier,
  LineComment,
  NumericLiteral,
  Whitespace,
  Newline,

  // Keyword,
  Enum,
  Int,
  Char,
  If,
  Else,
  While,
  Return,

  // Operator
  Sizeof,
  Negate,
  Minus,
  MinusAssign,
  Plus,
  PlusAssign,
  Star,
  StarStar,
  StarAssign,
  Div,
  DivAssign,
  Tilde,
  And,
  AndAnd,
  Or,
  OrOr,
  Xor,
  Increment,
  Decrement,
  Modulus,
  ModulusAssign,
  Assign,
  Equal,
  NotEqual,
  Greater,
  GreaterEqual,
  Less,
  LessEqual,
  RightShift,
  LeftShift,

  // Punctuation
  LeftParen,
  RightParen,
  LeftSquare,
  RightSquare,
  LeftBracket,
  RightBracket,
  Question,
  Colon,
  Comma,
  SemiColon,
  Dot,
}

export const TokenEOF = {
  type: TokenType.EOF,
  source: '',
}

export interface Token {
  type: TokenType
  source: string
}

// TODO: refactor as class for simple construction
export interface StringLiteral extends Token {
  value: string
}

export interface Whitespace extends Token {
  value: string
}

export interface Newline extends Token {
  value: string
}

export interface Identifier extends Token {
  name: string
}

export interface LineComment extends Token {
  value: string
}

export interface NumericLiteral extends Token {
  value: number
}
