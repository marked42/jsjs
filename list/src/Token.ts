export class Token {
  constructor(public type: TokenType, public text: string) {}
}

export enum TokenType {
  Invalid = 'invalid',
  EOF = 'eof',
  Name = 'name',
  Comma = 'comma',
  Equal = 'equal',
  LeftBracket = 'leftBracket',
  RightBracket = 'rightBracket',
}
