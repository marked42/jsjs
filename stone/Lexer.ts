import {
  EOF,
  IdentifierToken,
  NumberToken,
  PunctuatorToken,
  StringToken,
  Token,
  TokenStream,
  WhitespaceToken,
} from './Token'

const CHAR_EOF = ''

export class Lexer extends TokenStream {
  private tokens: Token[] = []

  private inputIndex = 0
  private inputRow = 0
  private inputCol = 0

  constructor(private input: string) {
    super()
  }

  peek(i: number = 0) {
    this.fillQueue(i)
    if (i < this.tokens.length) {
      return this.tokens[i]
    }
    return EOF
  }

  next() {
    const token = this.peek()
    if (token !== EOF) {
      this.tokens.shift()
    }
    return token
  }

  getTokens() {
    const tokens: Token[] = []
    while (true) {
      const token = this.next()
      if (token === EOF) {
        break
      }
      tokens.push(token)
    }
    return tokens
  }

  private peekChar(i = 0) {
    const index = this.inputIndex + i
    if (index >= this.input.length) {
      return CHAR_EOF
    }

    const char = this.input[index]

    return char
  }

  private nextChar() {
    const char = this.peekChar()
    if (char === CHAR_EOF) {
      return char
    }

    if (
      char === '\r' ||
      (char === '\n' && this.input[this.inputIndex - 1] !== '\r')
    ) {
      this.inputRow++
      this.inputCol = 0
    } else {
      this.inputCol++
    }

    this.inputIndex++
    return char
  }

  /**
   * 向缓冲队列中读入i个Token供消费
   */
  private fillQueue(i: number) {
    loop: while (true) {
      const notEnoughTokens = this.tokens.length <= i
      const hasMoreInput = this.inputIndex < this.input.length

      if (!hasMoreInput || !notEnoughTokens) {
        break
      }

      const start = {
        row: this.inputRow,
        col: this.inputCol,
      }
      const from = this.inputIndex
      const matches = [
        {
          test: this.isDigit,
          type: NumberToken,
          match: this.matchNumber,
        },
        {
          test: (char: string) => /[_a-zA-Z]/.test(char),
          type: IdentifierToken,
          match: this.matchIdentifier,
        },
        {
          test: (char: string) => char === '"',
          type: StringToken,
          match: this.matchString,
        },
        {
          test: (char: string) => /\p{Punct}/.test(char),
          type: PunctuatorToken,
          match: this.matchPunctuator,
        },
        {
          test: (char: string) => /\s/.test(char),
          type: WhitespaceToken,
          match: this.matchWhitespace,
        },
      ]
      // if (this.peekChar() === CHAR_EOF) {
      //   this.tokens.push(EOF)
      //   break
      // }

      for (const option of matches) {
        const { test, type, match } = option
        if (test(this.peekChar())) {
          this.tokens.push(
            // @ts-ignore
            new type(match.call(this), {
              start,
              end: {
                row: this.inputRow,
                col: this.inputCol,
              },
              from,
              to: this.inputIndex,
            })
          )
          continue loop
        }
      }

      this.throwTokenError('unknown type')
    }
  }

  isDigit(char: string) {
    if (char === CHAR_EOF) {
      return false
    }

    const min = '0'.codePointAt(0)
    const max = '9'.codePointAt(0)
    const codePoint = char.codePointAt(0)

    if (min === void 0 || max === void 0 || codePoint === void 0) {
      throw new Error('received code point undefined for character ' + char)
    }
    return min <= codePoint && codePoint <= max
  }

  isNonZeroDigit(char: string) {
    const min = '1'.codePointAt(0)
    const max = '9'.codePointAt(0)
    const codePoint = char.codePointAt(0)

    if (char === CHAR_EOF) {
      return true
    }

    if (min === void 0 || max === void 0 || codePoint === void 0) {
      throw new Error('received code point undefined for character ' + char)
    }
    return min <= codePoint && codePoint <= max
  }

  private matchNumber() {
    const char = this.peekChar()
    if (char === '0') {
      this.nextChar()
      if (this.isDigit(this.peekChar())) {
        throw new Error('no digits allowed after zeros, input ' + this.input)
      } else {
        return '0'
      }
    } else if (this.isNonZeroDigit(char)) {
      this.nextChar()
      let text = char
      while (this.isDigit(this.peekChar())) {
        text += this.nextChar()
      }
      return text
    }

    this.throwTokenError('number')
  }

  private matchIdentifier() {
    const char = this.peekChar()

    let text = ''
    if (/^[_a-zA-Z]$/.test(char)) {
      text += this.nextChar()
      while (/^[_a-zA-Z0-9]$/.test(this.peekChar())) {
        text += this.nextChar()
      }
      return text
    }

    this.throwTokenError('identifier')
  }

  throwTokenError(tokenType: string) {
    throw new Error(
      `expect ${tokenType} token at index ${this.inputIndex}, position: (${this.inputRow}, ${this.inputCol})`
    )
  }

  private matchWhitespace() {
    const char = this.peekChar()
    if (!/[ \t\n\r]/.test(char)) {
      this.throwTokenError('whitespace')
    }

    let text = ''
    while (/[ \t\n\r]/.test(this.peekChar())) {
      text += this.nextChar()
    }

    return text
  }

  private matchString() {
    let text = ''
    this.match('"')
    if (this.peekChar() === '\\') {
      text += this.nextChar()

      if (this.match(/(n|r|v|t)/)) {
        text += this.nextChar()
      } else {
        text += this.nextChar()
      }
    } else if (this.match(/[^\n\r\\"]/)) {
      while (this.match(/[^\n\r\\"]/)) {
        text += this.nextChar()
      }
    } else {
      this.throwTokenError('string')
    }

    this.match('"')
    return text
  }

  match(expected: string | RegExp | ((char: string) => boolean)) {
    const char = this.peekChar()

    if (
      (typeof expected === 'string' && expected === char) ||
      (expected instanceof RegExp && expected.test(char)) ||
      (typeof expected === 'function' && expected(char))
    ) {
      return this.nextChar()
    }

    throw new Error(
      'failed to match character, expect ${expected}, got ${char}'
    )
  }

  private matchPunctuator() {
    if (this.match(/\p{Punct}/)) {
      return this.nextChar()
    }
    this.throwTokenError('punctuator')
  }
}
