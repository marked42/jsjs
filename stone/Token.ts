export enum TokenType {
  String,
  Number,
  Identifier,
  Punctuator,
  Whitespace,
  EOF,
}

interface Position {
  row: number
  col: number
}

interface SourceLocation {
  start?: Position
  end?: Position
  from?: number
  to?: number
}

export abstract class Token {
  constructor(private _text: string, private _location?: SourceLocation) {}
  abstract type(): TokenType
  text() {
    return this._text
  }
  location(): SourceLocation {
    if (this._location !== void 0) {
      return this._location
    }
    return {
      start: void 0,
      end: void 0,
      from: void 0,
      to: void 0,
    }
  }
}

export class NumberToken extends Token {
  type() {
    return TokenType.Number
  }
}

export class StringToken extends Token {
  type() {
    return TokenType.String
  }
}

export class PunctuatorToken extends Token {
  type() {
    return TokenType.Punctuator
  }
}

export class IdentifierToken extends Token {
  type() {
    return TokenType.Identifier
  }
}

export class WhitespaceToken extends Token {
  type() {
    return TokenType.Whitespace
  }
}

export const EOF = (function () {
  class EOFToken extends Token {
    type() {
      return TokenType.EOF
    }
  }

  return new EOFToken('')
})()

export abstract class TokenStream {
  abstract peek(i: number): Token
  abstract peek(): Token

  abstract next(): Token
  end() {
    return this.peek() === EOF
  }
}
