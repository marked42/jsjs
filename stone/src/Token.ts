export enum TokenType {
  String,
  Number,
  Identifier,
  Punctuator,
  Whitespace,
  EOF,
}

interface Position {
  row: number
  col: number
}

interface SourceLocation {
  start?: Position
  end?: Position
  from?: number
  to?: number
}

export abstract class Token {
  constructor(private text: string, private location?: SourceLocation) {}
  abstract type(): TokenType
  getText() {
    return this.text
  }
  getLocation(): SourceLocation {
    if (this.location !== void 0) {
      return this.location
    }
    return {
      start: void 0,
      end: void 0,
      from: void 0,
      to: void 0,
    }
  }
}

export class NumberToken extends Token {
  type() {
    return TokenType.Number
  }
}

export class StringToken extends Token {
  type() {
    return TokenType.String
  }
}

export class PunctuatorToken extends Token {
  type() {
    return TokenType.Punctuator
  }
}

export class IdentifierToken extends Token {
  type() {
    return TokenType.Identifier
  }
}

export class WhitespaceToken extends Token {
  type() {
    return TokenType.Whitespace
  }
}

export const EOF = (function () {
  class EOFToken extends Token {
    type() {
      return TokenType.EOF
    }
  }

  return new EOFToken('')
})()

export abstract class TokenStream {
  abstract peek(i: number): Token
  abstract peek(): Token

  abstract next(): Token
  end() {
    return this.peek() === EOF
  }
}
