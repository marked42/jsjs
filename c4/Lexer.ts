import { CharacterStream } from './CharacterStream'
import { Token, TokenType, Identifier, StringLiteral } from './Token'
import * as Errors from './Errors'

export class Lexer {
  private codes: number[] = []
  private charCode: number = -1
  constructor(private stream: CharacterStream) {}

  getChar() {
    this.charCode = this.stream.next()
    if (this.charCode !== CharacterStream.EOF) {
      this.codes.push(this.charCode)
    }
    return this.charCode
  }

  ungetChar() {
    this.codes.splice(this.codes.length - 1, 1)
    this.stream.prev()
    this.charCode = this.codes[this.codes.length - 1]
  }

  resetCharCodes() {
    this.codes = []
    this.charCode = -1
  }

  next(): Token {
    this.getChar()
    // StringLiteral
    if (charCodeIs(this.charCode, '"')) {
      const value = []

      while (true) {
        this.getChar()

        if (charCodeIs(this.charCode, '\\')) {
          this.getChar()
          if (charCodeIs(this.charCode, '"')) {
            value.push('"'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, "'")) {
            value.push("'".charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'b')) {
            value.push('\b'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'f')) {
            value.push('\f'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'n')) {
            value.push('\n'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'r')) {
            value.push('\r'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 't')) {
            value.push('\t'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'v')) {
            value.push('\v'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, '\\')) {
            value.push('\\'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, '\r')) {
            value.push('\r'.charCodeAt(0))
            this.getChar()
            if (charCodeIs(this.charCode, '\n')) {
              value.push('\n'.charCodeAt(0))
            } else {
              this.ungetChar()
            }
          } else if (charCodeIs(this.charCode, '\n')) {
            value.push('\n'.charCodeAt(0))
          } else {
            console.error(
              'invalid escape sequence \\' + String.fromCharCode(this.charCode)
            )
            value.push(this.charCode)
          }
        } else if (
          charCodeIs(this.charCode, '\r') ||
          charCodeIs(this.charCode, '\n')
        ) {
          throw Errors.StringLiteralUnescapedNewline
        } else if (charCodeIs(this.charCode, '"')) {
          const source = String.fromCharCode(...this.codes)
          this.resetCharCodes()
          return {
            type: TokenType.StringLiteral,
            source,
            value: String.fromCharCode(...value),
          } as StringLiteral
        } else if (this.charCode === CharacterStream.EOF) {
          throw Errors.StringLiteralEndEarly
        } else {
          value.push(this.charCode)
        }
      }
      // NumericLiteral
    } else if (isDigit(this.charCode)) {
      // LineComment
    } else if (charCodeIs(this.charCode, '/')) {
      this.stream.next()

      if (charCodeIs(this.stream.peek(), '/')) {
        this.stream.next()
        const codes = []
        while (true) {
          const code = this.stream.peek()
          if (!charCodeIs(code, '\r') && !charCodeIs(code, '\n')) {
            codes.push(code)
            this.stream.next()
          } else {
            return {
              type: TokenType.LineComment,
              source: String.fromCharCode(...codes),
            }
          }
        }
      } else {
        return {
          type: TokenType.Operator,
          source: '/',
        }
      }
      // Whitespace
    } else if (isWhitespace(this.charCode)) {
      const codes = [this.stream.next()]
      while (isWhitespace(this.stream.peek())) {
        codes.push(this.stream.next())
      }

      return {
        type: TokenType.Whitespace,
        source: String.fromCharCode(...codes),
      }
      // Newline
    } else if (charCodeIs(this.charCode, '\r')) {
      this.stream.next()
      if (charCodeIs(this.stream.peek(), '\n')) {
        this.stream.next()
        return {
          type: TokenType.Newline,
          source: '\r\n',
        }
      } else {
        return {
          type: TokenType.Newline,
          source: '\r',
        }
      }
    } else if (charCodeIs(this.charCode, '\n')) {
      this.stream.next()
      return {
        type: TokenType.Whitespace,
        source: '\n',
      }
    } else if (isIdentifierStart(this.charCode)) {
      const codes = [this.stream.next()]

      while (true) {
        const code = this.stream.peek()
        if (isIdentifierContinue(code)) {
          codes.push(code)
          this.stream.next()
          continue
        }

        const source = String.fromCharCode(...codes)
        return {
          type: TokenType.Identifier,
          source,
          name: source,
        } as Identifier
      }
    }

    throw new Error('Unexpected token ' + String.fromCharCode(this.charCode))
  }
}

function charCodeIs(code: number, text: string) {
  if (text.length !== 1) {
    throw new Error('text must be single character, got ' + text)
  }
  return code === text.charCodeAt(0)
}

function isDigit(code: number) {
  return code >= '0'.charCodeAt(0) && code <= '9'.charCodeAt(0)
}

function isWhitespace(code: number) {
  const ws = [' ', '\t'].map((c) => c.charCodeAt(0))

  return ws.includes(code)
}

function isLetter(code: number) {
  const isLower = code >= 'a'.charCodeAt(0) && code <= 'z'.charCodeAt(0)
  const isUpper = code >= 'A'.charCodeAt(0) && code <= 'Z'.charCodeAt(0)

  return isLower || isUpper
}

function isIdentifierStart(code: number) {
  return isLetter(code) || charCodeIs(code, '_')
}

function isIdentifierContinue(code: number) {
  return isLetter(code) || charCodeIs(code, '_') || isDigit(code)
}
