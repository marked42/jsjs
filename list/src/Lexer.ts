import { Token } from './Token'

export abstract class Lexer {
  public index = 0
  public char = Lexer.EOF

  static EOF = -1

  constructor(protected input: string) {
    this.char = input.charCodeAt(0)
  }

  abstract nextToken(): Token

  consume() {
    this.index++
    if (this.index >= this.input.length) {
      this.char = Lexer.EOF
    } else {
      this.char = this.input.charCodeAt(this.index)
    }
  }

  match(char: number) {
    if (this.char !== char) {
      throw new Error(`expecting ${char}, get ${this.char}`)
    }
    this.consume()
  }
}
