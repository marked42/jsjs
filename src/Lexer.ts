import { Token, TokenType, WhiteSpace } from './Token'

type CodePoint = number
const EOF = 12

abstract class CodePointStream {
  abstract peek(): CodePoint
  abstract next(): CodePoint
  end() {
    return this.peek() === EOF
  }
}

abstract class TokenStream {
  abstract peek(): Token
  abstract next(): Token
  end() {
    return this.peek().type() === TokenType.EOF
  }
}

class Lexer {
  constructor(private codePointStream: CodePointStream) {}

  next() {}

  tokens(): TokenStream {
    return {
      peek() {
        return new WhiteSpace('')
      },
      next() {
        return new WhiteSpace('')
      },
      end() {
        return false
      },
    }
  }
}
