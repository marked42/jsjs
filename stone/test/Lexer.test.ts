import { Lexer } from '../Lexer'
import { NumberToken } from '../Token'
// describe('Lexer', () => {
//   it('should lex number', () => {
//     const input = '111'
//     const lexer = new Lexer(input)

//     const tokens = lexer.getTokens()
//     console.log('tokens: ', lexer.getTokens())
//     // expect(tokens).toEqual([new NumberToken('111')])
//   })

//   it('should lex number', () => {
//     const input = '0'
//     const lexer = new Lexer(input)

//     const tokens = lexer.getTokens()
//     console.log('tokens: ', lexer.getTokens())
//     // expect(tokens).toEqual([new NumberToken('111')])
//   })

//   //   it('should lex number', () => {
//   //     const input = '00'
//   //     const lexer = new Lexer(input)

//   //     const tokens = lexer.getTokens()
//   //     console.log('tokens: ', lexer.getTokens())
//   //     // expect(tokens).toEqual([new NumberToken('111')])
//   //   })
// })

const input = '012'
const lexer = new Lexer(input)

console.log('tokens: ', lexer.getTokens())
