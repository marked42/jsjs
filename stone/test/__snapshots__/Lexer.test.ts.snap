// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compound example should parse multiple tokens 1`] = `
Array [
  IdentifierToken {
    "location": Object {
      "end": Object {
        "col": 3,
        "row": 0,
      },
      "from": 0,
      "start": Object {
        "col": 0,
        "row": 0,
      },
      "to": 3,
    },
    "text": "let",
  },
  IdentifierToken {
    "location": Object {
      "end": Object {
        "col": 5,
        "row": 0,
      },
      "from": 4,
      "start": Object {
        "col": 4,
        "row": 0,
      },
      "to": 5,
    },
    "text": "a",
  },
  PunctuatorToken {
    "location": Object {
      "end": Object {
        "col": 7,
        "row": 0,
      },
      "from": 6,
      "start": Object {
        "col": 6,
        "row": 0,
      },
      "to": 7,
    },
    "text": "=",
  },
  NumberToken {
    "location": Object {
      "end": Object {
        "col": 9,
        "row": 0,
      },
      "from": 8,
      "start": Object {
        "col": 8,
        "row": 0,
      },
      "to": 9,
    },
    "text": "1",
  },
  PunctuatorToken {
    "location": Object {
      "end": Object {
        "col": 10,
        "row": 0,
      },
      "from": 9,
      "start": Object {
        "col": 9,
        "row": 0,
      },
      "to": 10,
    },
    "text": ",",
  },
  IdentifierToken {
    "location": Object {
      "end": Object {
        "col": 12,
        "row": 0,
      },
      "from": 11,
      "start": Object {
        "col": 11,
        "row": 0,
      },
      "to": 12,
    },
    "text": "b",
  },
  PunctuatorToken {
    "location": Object {
      "end": Object {
        "col": 14,
        "row": 0,
      },
      "from": 13,
      "start": Object {
        "col": 13,
        "row": 0,
      },
      "to": 14,
    },
    "text": "=",
  },
  StringToken {
    "location": Object {
      "end": Object {
        "col": 20,
        "row": 0,
      },
      "from": 15,
      "start": Object {
        "col": 15,
        "row": 0,
      },
      "to": 20,
    },
    "text": "\\"str\\"",
  },
]
`;
